--- a/mozilla/security/nss/cmd/modutil/install-ds.c
+++ b/mozilla/security/nss/cmd/modutil/install-ds.c
@@ -252,7 +252,7 @@ Pk11Install_File_Generate(Pk11Install_Fi
 					goto loser;
 				}
 				_this->permissions = (int) strtol(subval->string, &endp, 8);
-				if(*endp != '\0' || subval->string == "\0") {
+				if(*endp != '\0' || subval->string[0] == '\0') {
 					errStr = PR_smprintf(errString[BOGUS_FILE_PERMISSIONS],
                                     _this->jarPath);
 					goto loser;
